stock_picking_product_interchangeable/models/stock_move.py

```diff
@@ -0,0 +1,75 @@
+from odoo import models
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    def _prepare_interchangeable_products(self, mode, for_qty):
+        """
+        Preparing products for current product replacement
+        :param move_id: stock.move record
+        :param mode: stock.picking.type substitute_products_mode value
+        :param for_qty: product qty for replacement
+        :return: list of tuple [(product_obj, count), ...]
+        """
+        qty = for_qty
+        if qty >= 0:
+            return False, False
+        products = []
+        interchangeable_products_qty = 0
+        for product in self.product_id.product_interchangeable_ids:
+            available_qty = product.immediately_usable_qty
+            if available_qty > 0 > qty:
+                product_qty = abs(qty) if available_qty + qty >= 0 else available_qty
+                products.append((product, product_qty))
+                qty += product_qty
+                interchangeable_products_qty += product_qty
+        if (mode == "all" and qty == 0) or mode == "any":
+            return products, interchangeable_products_qty
+        return False, False
+
+    def _create_stock_move_interchangeable_products(self, products):
+        """
+        Creates stok.move records for replacement product
+        :param move_id: parent stock.move record
+        :param products: struct list of tuple [(product_obj, count), ...]
+        :return: None
+        """
+        moves = self.env["stock.move"]
+        stock_move_obj = self.env["stock.move"]
+        if not products:
+            return moves
+        return stock_move_obj.create(
+            [
+                {
+                    "picking_id": self.picking_id.id,
+                    "name": product.display_name,
+                    "product_id": product.id,
+                    "product_uom_qty": qty,
+                    "location_id": self.location_id.id,
+                    "location_dest_id": self.location_dest_id.id,
+                    "company_id": self.company_id.id,
+                }
+                for product, qty in products
+            ]
+        )
+
+    def _action_confirm(self, merge=True, merge_into=False):
+        moves = super(StockMove, self)._action_confirm(
+            merge=merge, merge_into=merge_into
+        )
+        type_ = moves.picking_type_id
+        mode = type_.substitute_products_mode
+        skip_behavior = not (mode and type_.code == "outgoing")
+        if skip_behavior or moves.picking_id.pass_interchangeable:
+            return moves
+        move_ids = moves.filtered(lambda m: m.product_id.product_interchangeable_ids)
+        new_moves = self.env["stock.move"]
+        for move in move_ids:
+            qty = move.product_id.immediately_usable_qty
+            products, products_qty = move._prepare_interchangeable_products(mode, qty)
```

Comment:
can we calc `interchangeable_products_qty` as sum(product_info[1] for product_info in products_info) and thereby simplify the method `_prepare_interchangeable_products` it seems that this method, in addition to preparing data for a particular product, should not do summation, which is used only in one place, what do you think?

Reply:
I suggest renaming the variable that assigns the value to the `_prepare_interchangeable_products` method, since it does not return a recordset, but information about the product and its quantity. for example:
`products_info = move._prepare_interchangeable_products(mode, qty)`

Reply:
Resolved